File: samples/replicatedentity-counter/src/main/java/com/example/counter/domain/Counter.java
package com.example.counter.domain;

import com.akkaserverless.javasdk.replicatedentity.ReplicatedCounter;
import com.akkaserverless.javasdk.replicatedentity.ReplicatedEntityContext;
import com.example.counter.CounterApi;
import com.google.protobuf.Empty;

public class Counter extends AbstractCounter {
  @SuppressWarnings("unused")
  private final String entityId;

  public Counter(ReplicatedEntityContext context) {
    this.entityId = context.entityId();
  }

  @Override
  public Effect<Empty> increase(ReplicatedCounter counter, CounterApi.IncreaseValue command) {
    if (command.getValue() < 0) {
      return effects().error("Increase requires a positive value. It was [" + command.getValue() + "].");
    }

    counter.increment(command.getValue());

    return effects().update(counter).thenReply(Empty.getDefaultInstance());
  }

  @Override
  public Effect<Empty> decrease(ReplicatedCounter counter, CounterApi.DecreaseValue command) {
    if (command.getValue() < 0) {
      return effects().error("Decrease requires a positive value. It was [" + command.getValue() + "].");
    }

    counter.decrement(command.getValue());

    return effects().update(counter).thenReply(Empty.getDefaultInstance());
  }

  @Override
  public Effect<CounterApi.CurrentCounter> getCurrentCounter(ReplicatedCounter counter, CounterApi.GetCounter command) {
    CounterApi.CurrentCounter current = CounterApi.CurrentCounter.newBuilder().setValue(counter.getValue()).build();
    return effects().reply(current);
  }
}

File: samples/scala-replicated-examples/src/main/scala/com/example/replicated/counter/Counter.scala
package com.example.replicated.counter
import com.akkaserverless.scalasdk.replicatedentity.ReplicatedCounter
import com.akkaserverless.scalasdk.replicatedentity.ReplicatedEntity
import com.akkaserverless.scalasdk.replicatedentity.ReplicatedEntityContext
import com.google.protobuf.empty.Empty

class Counter(context: ReplicatedEntityContext) extends AbstractCounter {

  override def increase(counter: ReplicatedCounter, increaseValue: IncreaseValue): ReplicatedEntity.Effect[Empty] =
    effects
      .update(counter.increment(increaseValue.value))
      .thenReply(Empty.defaultInstance)

  override def decrease(counter: ReplicatedCounter, decreaseValue: DecreaseValue): ReplicatedEntity.Effect[Empty] =
    effects
      .update(counter.decrement(decreaseValue.value))
      .thenReply(Empty.defaultInstance)

  override def get(counter: ReplicatedCounter, getValue: GetValue): ReplicatedEntity.Effect[CurrentValue] =
    effects.reply(CurrentValue(counter.value))

}
